src/map/script.c

```diff
@@ -17084,6 +17084,1007 @@ BUILDIN(getunittype) {
 	return true;
 }
 
+/**
+ * Sets real-time unit data for a game object.
+ * Setunitdata <GUID>,<DataType>,<Val1>{,<Val2>,<Val3>}
+ * @param1  GUID        Game ID of the unit.
+ * @param2  DataType    Type of Data to be set for the unit.
+ * @param3  Value#1     Value to be passed as change in data.
+ * @param4  Value#2     Optional int value to be passed for certain data types.
+ * @param5  Value#3     Optional int value to be passed for certain data types.
+ * @return  1 on success, 0 on failure.
+ */
+BUILDIN(setunitdata)
+{
+	struct block_list *bl = NULL;
+	const char *mapname = NULL, *udtype = NULL;
+	int type = 0, value = 0, value2 = 0, value3 = 0;
+	
+	bl = map->id2bl(script_getnum(st, 2));
+	
+	if (bl == NULL) {
+		ShowWarning("buildin_setunitdata: Error in finding object with given GID %d!\n", script_getnum(st, 2));
+		script_pushint(st, 0);
+		return false;
+	} else if (!script_isstring(st, 3) || (udtype = script_getstr(st, 3)) == NULL || !script->get_constant(udtype, &type)) {
+		ShowWarning("buildin_setunitdata: Undefined Unit Data Type %s set.\n", udtype);
+		script_pushint(st, 0);
+		return false;
+	} else if (!script_hasdata(st, 4)) {
+		ShowWarning("buildin_setunitdata: Value not given for type %s.\n", udtype);
+		script_pushint(st, 0);
+		return false;
+	}
+	
+	/* Mandatory Value1 */
+	if (script_isstring(st, 4)) {
+		if (type == UDT_MAPIDXY) {
+			mapname = script_getstr(st, 4);
+		} else if (!script->get_constant(script_getstr(st, 4), &value)) {
+			ShowError("buildin_setunitdata: Invalid data type for Parameter");
+			script_pushint(st, 0);
+			return false;
+		}
+	} else if (script_isint(st, 4)) {
+		value = script_getnum(st, 4);
+	} else {
+		ShowError("buildin_setunitdata: Invalid data type for Parameter #4");
+		script_pushint(st, 0);
+		return false;
+	}
+	
+	/* Optional Value2 */
+	if (script_hasdata(st, 5)) {
+		if (script_isstring(st, 5)) {
+			ShowError("buildin_setunitdata: Fifth parameter expects integer, string given.");
+			script_pushint(st, 0);
+			return false;
+		} else {
+			value2 = script_getnum(st, 5);
+		}
+	}
+	
+	/* Optional Value3 */
+	if (script_hasdata(st, 6)) {
+		if (script_isstring(st, 6)) {
+			ShowError("buildin_setunitdata: Fifth parameter expects integer, string given.");
+			script_pushint(st, 0);
+			return false;
+		} else {
+			value3 = script_getnum(st, 6);
+		}
+	}
+	
+	switch (bl->type) {
```

Comment:
Sorry my screen is 27" wide. Though none of the statements are long enough to cause reading problems.